internal/resources/scope.go

```diff
@@ -0,0 +1,347 @@
+	" state of the scope. Please run `terraform plan` after 1-2 minutes to know the" +
+	" current scope state. Additionally, run `terraform apply --refresh-only` to update" +
+	" the state from remote, unexpected error: "
+
+// Scope is the scope resource implementation.
+type Scope struct {
+	*providerschema.Data
+}
+
+// NewScope is a helper function to simplify the provider implementation.
+func NewScope() resource.Resource {
+	return &Scope{}
+}
+
+func (s *Scope) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
+	resource.ImportStatePassthroughID(ctx, path.Root("name"), req, resp)
+}
+
+// Metadata returns the Scope resource type name.
+func (s *Scope) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_scope"
+}
+
+// Schema defines the schema for the Scope resource.
+func (s *Scope) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = ScopeSchema()
+}
+
+// Configure It adds the provider configured api to the scope resource.
+func (s *Scope) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	data, ok := req.ProviderData.(*providerschema.Data)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *ProviderSourceData, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+
+		return
+	}
+
+	s.Data = data
+}
+
+func (s *Scope) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan providerschema.Scope
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	ScopeRequest := scope_api.CreateScopeRequest{
+		Name: plan.Name.ValueString(),
+	}
+
+	if err := s.validateCreateScopeRequest(plan); err != nil {
+		resp.Diagnostics.AddError(
+			"Error parsing create scope request",
+			"Could not create scope, unexpected error: "+err.Error(),
+		)
+		return
+	}
+
+	var organizationId = plan.OrganizationId.ValueString()
+	var projectId = plan.ProjectId.ValueString()
+	var clusterId = plan.ClusterId.ValueString()
+	var bucketId = plan.BucketId.ValueString()
+
+	url := fmt.Sprintf("%s/v4/organizations/%s/projects/%s/clusters/%s/buckets/%s/scopes", s.HostURL, organizationId, projectId, clusterId, bucketId)
+	cfg := api.EndpointCfg{Url: url, Method: http.MethodPost, SuccessStatus: http.StatusCreated}
+	_, err := s.Client.ExecuteWithRetry(
+		ctx,
+		cfg,
+		ScopeRequest,
+		s.Token,
+		nil,
+	)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error creating scope",
+			errorMessageWhileScopeCreation+api.ParseError(err),
+		)
+		return
+	}
+
+	diags = resp.State.Set(ctx, plan)
```

Comment:
I believe we should not directly set the plan in the state file. Instead, we should populate the computed field of the plan before executing the set operation, similar to how 'initializeApiKeyWithPlanAndId' works in the apikey resource

Reply:
@aniket-Kumar-c but the scope doesn't return a response, also only the name is a required field, collections and uid are neither required, nor optional, they are just populated from remote. So how to create func  initializeScopeWithPlanAndName? Something like this?
```
//func initializeScopeWithPlanAndName(plan providerschema.Scope, name string) providerschema.Scope {
//	plan.Name = types.StringValue(name)
//      plan.Collections = types.SetNull() - not sure how to set the collections to null, as it is of types.Set
//	plan.Uid = types.StringNull()
//	return plan
//}
```

Reply:
Yes, Before setting the computed field to the state file, we need to ensure it's null. To verify the effectiveness of this operation, intentionally introduce an error during the get operation. Afterward, perform the plan again; it should successfully read from the remote server. The plan results should indicate that only the computed field has been modified.

Reply:
@aniket-Kumar-c I tested this scenario on the current implementation-
1. Purposely introduced changes to the computed values (scope uid, collection name, uid, maxTTL) in the state file.
2. Ran terraform plan, successfully read from the remote server and indicated that computed fields have been modified-
<img width="1386" alt="Screen Shot 2024-03-06 at 11 05 28 PM" src="https://github.com/couchbasecloud/terraform-provider-couchbase-capella/assets/124640240/b766d1f1-103a-4e5d-9542-6e9f5be5e7e2">
<img width="997" alt="Screen Shot 2024-03-06 at 11 05 40 PM" src="https://github.com/couchbasecloud/terraform-provider-couchbase-capella/assets/124640240/82419b26-8ce8-4f4c-899e-2b077306aa79">
So this seems to be working even now? Is my understanding correct?🤔